src/int_allocator.rs

```diff
@@ -6,17 +6,25 @@ pub struct IntAtomBuf {
     end: u32,
 }
 
-enum NodePtr {
-    Atom(IntAtomBuf),
-    Pair(u32, u32),
+#[derive(Clone, Copy)]
+pub struct IntPair {
+    first: i32,
+    rest: i32,
 }
 
 pub struct IntAllocator {
     // this is effectively a grow-only stack where atoms are allocated. Atoms
     // are immutable, so once they are created, they will stay around until the
     // program completes
     u8_vec: Vec<u8>,
-    node_vec: Vec<NodePtr>,
+
+    // storage for all pairs (positive indices)
+    pair_vec: Vec<IntPair>,
+
+    // storage for all atomes (negative indices).
```

Comment:
> atomes

Reply:
I'll force-push an update to fix this in a bit